### ISSUE TITLE
svg re-renders preview extra times

### ISSUE THREAD

**Issue by andykais** [ranger/ranger#1320]
#### Runtime Environment
```
 $ ranger --version
ranger version: ranger-master 1.9.2
Python version: 3.7.0 (default, Jul 15 2018, 10:44:58) [GCC 8.1.1 20180531]
Locale: en_US.UTF-8
```

#### Current Behavior
when displaying the preview of an svg, ranger renders a preview multiple times. Its hard to see in this gif here, but when it attempts an svg preview, it flashes after each preview has completed rendering. I also am hitting `"w"` to show the tasks being done when an svg is attempting a preview.
![ranger](https://user-images.githubusercontent.com/6627147/46309786-7478cd00-c58b-11e8-87b9-add5d09b40fe.gif)
This does not appear to happen with any of the other image previews (`png`, `jpg`, `pdf`)

#### Expected Behavior
svg only renders a preview once.

#### Possible Solutions
I do not know the internals of ranger, but I can confirm that the issue appears to be happening before `scope.sh` is being called. I know this because I put a log in `scope.sh` to see how many times it was being called for each file preview, and for svgs it was more than once.

#### Steps to reproduce
I have uncommented the lines in `scope.sh` that allow for svg previews
![screenshot 147](https://user-images.githubusercontent.com/6627147/46309430-5363ac80-c58a-11e8-882f-653ca1446b41.png)
uncomment the relevant lines in `scope.sh` and look at an svg preview.

**Comment by LexTek** [comment]
Hello, I encountered a similar problem, and fixed it by replacing the line
```sh
convert "${FILE_PATH}" "${IMAGE_CACHE_PATH}" && exit 6
```
by
```sh
convert -- "${FILE_PATH}" "${IMAGE_CACHE_PATH}" && exit 6
```
Does it works for you ?

**Comment by prosoitos** [comment]
May not be related, but I could not get previews for .svg at all, but removing `+xml` in the scope.sh section for svg solved it for me.

```sh
        # SVG
	image/svg)
	    convert "${FILE_PATH}" "${IMAGE_CACHE_PATH}" && exit 6
	    exit 1;;
```

**Comment by toonn** [comment]
I think you could simply do `image/svg|image/svg+xml)`.

**Comment by prosoitos** [comment]
Indeed. Much less clumsy. Thank you @toonn :slightly_smiling_face:

**Comment by toonn** [comment]
@prosoitos, thanks for reporting this even though it's not related to this issue. I've decided to include the `image/svg` mime type in scope by default #1570.

**Comment by prosoitos** [comment]
Great!

**Comment by hfel** [comment]
Hello! I'm still experiencing the problem of multiple renders; @LexTek's solution (which is the default in `scope.sh` now) didn't work.

**Comment by szsdk** [comment]
Hi, I am also suffering from this strange bug. I found a solution which is using `inkscpae` to replace the `convert`.
```bash
         image/svg|image/svg+xml)
             #convert -- "${FILE_PATH}" "${IMAGE_CACHE_PATH}" && exit 6
             inkscape "${FILE_PATH}" -e "${IMAGE_CACHE_PATH}" && exit 6
             exit 1;;
```

**Comment by toonn** [comment]
@hfel, @szsdk, I have a suspicion. Could you test the following?
```
image/svg?(+xml))
    convert -- "${FILE_PATH}" "${IMAGE_CACHE_PATH}.tmp"
    mv "${IMAGE_CACHE_PATH}.tmp" "${IMAGE_CACHE_PATH}" && exit 6
    exit 1;;
```

**Comment by szsdk** [comment]
@toonn , the svg file is previewed in text with your code.
Then I changed it a little bit.
```
         image/svg+xml|image/svg)
            convert -- "${FILE_PATH}" "${IMAGE_CACHE_PATH}.tmp"
            mv "${IMAGE_CACHE_PATH}.tmp" "${IMAGE_CACHE_PATH}" && exit 6
            exit 1;;
```
Unfortrunately, it is not working. The problem is same.

**Comment by toonn** [comment]
Hmm, @szsdk, I suspect that's because `convert` is detecting the extension. Try flipping it around: `"tmp.${IMAGE_CACHE_PATH}"`.

**Comment by szsdk** [comment]
@tonttu  With `image/svg?(+xml))`, I still can not make it work in your new version. Only flip the order with the heading line I use, `image/svg+xml|image/svg)`, the problem is still there.

**Comment by toonn** [comment]
@tonttu, the case match doesn't really matter, having `tmp.` as a *pre*fix rather than a suffix is what matters, so the path still has a `.jpg` extension for `convert` to detect.

**Comment by szsdk** [comment]
@toonn Yeah, I tried both. Neither of them works.

**Comment by eMPee584** [comment]
.. the CommandLoader seems to get invoked multiple times in parallel, why is that :thinking:

**Comment by kankaristo** [comment]
I'm not sure if this is even related to the issue, but just in case someone finds this useful...

I tried @szsdk's Inkscape solution, but my version of Inkscape (from the Ubuntu repositories) doesn't have an `-e` flag.

But I did manage to get SVG previews working with Inkscape.

-----

I think my main issue with SVG previews was that the SVG files I was trying to preview had huge pixel resolutions. I like to use millimeters as the unit (instead of pixels), and I mostly use Inkscape to draw "blueprints" with a 1:1 scale. So, some of the SVGs I was previewing had a resolution of over 10000x10000.

`convert` was running for so long, that I thought it had just hung.  So, I had to limit the resolution of the converted file.

Another issue was that `IMAGE_CACHE_PATH` has a hardcoded .jpg extension, and Inkscape doesn't support exporting to JPEG.

I'm not sure if just changing and exporting `IMAGE_CACHE_PATH` would work with Ranger, if I want to use `.png` instead of `.jpg`?

As a workaround, I exported to `<filename>.jpg.png` and then just renamed the file back to `<filename>.jpg`, because the image previewer shouldn't rely on the extension to figure out the image type.

The below worked for me:
```bash
## SVG
image/svg+xml|image/svg)
    #convert -- "${FILE_PATH}" "${IMAGE_CACHE_PATH}" && exit 6
    inkscape --export-width="${DEFAULT_SIZE%x*}" "${FILE_PATH}" -o "${IMAGE_CACHE_PATH}.png" \
        && mv "${IMAGE_CACHE_PATH}.png" "${IMAGE_CACHE_PATH}" \
        && exit 6
    exit 1;;
```

-----

..But after doing that (and realizing that the issue was the huge resolution of my SVGs), I found a similar solution using `convert`:
```bash
#SVG
image/svg+xml|image/svg)
    #convert -- "${FILE_PATH}" "${IMAGE_CACHE_PATH}" && exit 6
    convert "${FILE_PATH}" -resize "${DEFAULT_SIZE%x*}" "${IMAGE_CACHE_PATH}" && exit 6
    exit 1;;
```

-----

I also set `DEFAULT_SIZE=800x450`, since my terminal is only 1080x640, and now all my previews are much faster (previews are less than 20 % of the original resolution, but still larger than the preview area in my terminal).

Changing `DEFAULT_SIZE` based on the terminal window's size might be the best solution, but I doubt there's a universal way of doing that.

**Comment by toonn** [comment]
@kankaristo, I can confirm you have to rename to `.jpg`, I don't like it much but it's hardcoded, something something w3mdisplay... Hysterical raisins.

`Scope.sh` actually sets `PV_WIDTH` and `PV_HEIGHT` which are the width and height of the preview pane in columns and rows respectively.

I'm gonna go ahead and implement your suggestion. I'm sticking to `DEFAULT_SIZE` because previews are cached by sixe and I think sticking to one size'll be more efficient cache-wise.

**Comment by toonn** [comment]
Testing of #2167 would be appreciated!

**Comment by kankaristo** [comment]
> `Scope.sh` actually sets `PV_WIDTH` and `PV_HEIGHT` which are the width and height of the preview pane in columns and rows respectively.

That's good to know, I might use those, since my terminal is always the same size.

-----

> Testing of #2167 would be appreciated!

It seems I didn't test that well enough... Testing again after clearing Ranger's image cache, and `convert` still hangs for big SVG files.

Running the `convert` command (outside of Ranger) and interrupting it with `Ctrl+C` shows an error about the maximum width or height being exceeded (but it seems to just freeze, and doesn't even show the error before interrupting):
```
convert-im6.q16: width or height exceeds limit `/tmp/magick-2023590cy2XpkXDVEnA' @ error/cache.c/OpenPixelCache/3911.
convert-im6.q16: no images defined `output.jpg' @ error/convert.c/ConvertImageCommand/3258.
```

`convert` may be trying to convert from SVG to JPEG first, and resize after that (which is a dumb thing to for SVGs, which could be converted to any target resolution)...

So, I'll switch back to my earlier Inkscape export command, it's a bit ugly because of the `.jpg.png` rename thing, but it works for all files I've tried.

I suppose the Ranger devs don't want to switch to Inkscape for SVG image previews? Not sure what Inkscape uses internally.

**Comment by toonn** [comment]
Since svg previews are commented out to begin with I don't mind adding inkscape. Ideally we'd get convert working though.

**Comment by kankaristo** [comment]
`convert` also seems to drop out most text in my SVGs even when the conversion succeeds. Based on reading some forum posts, ImageMagick seems to have pretty bad SVG support, since it's mostly a raster graphics program.

-----

I also tested `rsvg-convert` (from the `librsvg2-bin` package), and it worked fine. It might be a better option than Inkscape, since it has fewer dependencies (Inkscape is a GUI program, `rsvg-convert` is just for SVG conversion).

Depending on the file, it's also ~3.5 to ~6.5 times faster:
```
rsvg-convert input1.svg    0,10s user 0,01s system 97% cpu 0,110 total
inkscape input1.svg        0,35s user 0,09s system 105% cpu 0,421 total

rsvg-convert input2.svg    0,42s user 0,02s system 99% cpu 0,435 total
inkscape input2.svg        2,68s user 0,11s system 101% cpu 2,756 total
```

This is what I'm using now (and will probably stop experimenting for now...):
```bash
## SVG
image/svg+xml|image/svg)
    rsvg-convert --keep-aspect-ratio --width "${DEFAULT_SIZE%x*}" "${FILE_PATH}" -o "${IMAGE_CACHE_PATH}.png" \
        && mv "${IMAGE_CACHE_PATH}.png" "${IMAGE_CACHE_PATH}" \
        && exit 6
```

It has the same issue as Inkscape that it doesn't support converting to JPEG, so I'm doing the same rename thing.

**Comment by ice-pl** [comment]
in my case working with image/svg|image/svg+xml)

**Comment by toonn** [comment]
@kankaristo, would totally accept a PR adding that to scope.sh btw.

**Comment by kankaristo** [comment]
> @kankaristo, would totally accept a PR adding that to scope.sh btw.

I created a [PR for it](https://github.com/ranger/ranger/pull/2537).
